jupyterhub_singleuser_profiles/profiles.py

```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
       else:
         raise
 
+  def write_secret(self, secret_name, key_name, data):
+    secret = V1Secret()
+    secret.metadata = V1ObjectMeta(name=secret_name, labels={'app': 'jupyterhub'})
+    secret.string_data = {key_name: yaml.dump(data, default_flow_style=False)} #stringData instead of data here to make kubernetes parse this as a string without base64 encoding
+    try:
+      api_response = self.api_client.replace_namespaced_secret(secret_name, self.namespace, secret)
+    except ApiException as e:
+      if e.status == 404:
+        try:
+          api_response = self.api_client.create_namespaced_secret(self.namespace, secret)
+        except ApiException as e:
+          _LOGGER.error("Exception when calling CoreV1Api->create_namespaced_secret: %s\n" % e)
+      else:
+        raise
+
   def update_user_profile_cm(self, username, data={}):
     
     user_cm = self.get_user_profile_cm(username)
+    user_cm_env = []
+    user_secret = self.get_user_secret(username)
+    user_secret['env'] = []
     cm_name = _USER_CONFIG_MAP_TEMPLATE % escape(username)
     cm_key_name = "profile"
     for key, value in data.items():
       user_cm[key] = value
+    for var in user_cm['env']:
+      if var['type'] == "password":
+        user_secret['env'].append(var) 
+      else:
+        user_cm_env.append(var)
+    user_cm['env'] = user_cm_env
+    self.write_secret(cm_name, cm_key_name, user_secret) #not a typo, secret uses the same name as user cm
     self.write_config_map(cm_name, cm_key_name, user_cm)
 
   def get_user_profile_cm(self, username):
     cm = self.read_config_map(_USER_CONFIG_MAP_TEMPLATE % escape(username), "profile")
+    secret = self.get_user_secret(username)
     if cm == {}:
-      return _DEFAULT_USER_CM
+      cm = _DEFAULT_USER_CM
+    for sec in secret['env']:
+      cm['env'].append(sec)
+    return cm
+
+  def get_user_secret(self, username):
+    secret = self.read_secret(_USER_CONFIG_MAP_TEMPLATE % escape(username))
+    if secret == {}:
+      return{"env":[]}
     else:
-      return cm
+      return secret
+  
+  def update_user_secret(self, username, data={}):
```

Comment:
Where is this used?

Reply:
I decided not to use this method, since I never had a use case where I needed to change just a single secret value. I will delete it.

Reply:
I decided not to use this method, since I had no use cases where I needed to change some particular secrets in the secret ConfigMap. I will delete it.